### PR TITLE
WalletConnect transforms into Reown

### PR DESCRIPTION
**Pull request by creeppak**
## Summary

Replaced the WalletConnect integration with Reown. From UX perspective nothing really changed. If you look under the hood you'll notice it was updated both from Reown and our sides. This should improve stability, at least theoretically. We'll know this once we start testing it on different platforms and with different wallets.

This version is still likely to be unstable. It's already huge, so we're gonna fix bugs in a different PR.

## Known Issues

1. Wallet icons don't load.
2. New Branded GUI should be applied.
3. Can't switch to OP Sepolia Testnet when connected using MetaMask. Can switch to Mainnet and back to Sepolia however.
4. Some wallets doesn't support test networks. Add the Eth Mainnet to the chains settings and set is as Primary in order to connect to these of wallets.

### Fixed
1. Chain switching is broken.
2. System tries to open a deeplink for a new wallet even though was connected remotely using smartphone
_This creates an annoying bug, where file explorer opens up, whenever any input from user wallet is requested. What one has to do is just click on Unity editor to focus it._
3. The system tries to redirect user to a wallet whenever the session is being restored. This operation doesn't require any input from the user.
4. The system currently is not capable of identifying if a local or a remote wallet was connected.

## Potential improvements

- We should add a Reown configuration property for the redirection strategy, in case a remote wallet was connected and a confirmation is expected from the wallet app. Dev would want to notify user that he has to take action in his remote wallet to continue. We can show our info overlays by default.
- Ping wallet when restoring session (there is a command I believe) to make sure the session can be safely restored, behave like there's no active connection otherwise.

## Testing

All the platforms except WebGL should be tested to see how new integration behaves in different environments.

- [ ] Desktop/Editor
- [ ] Android
- [ ] iOS

WebGL is not supported currently as it uses another approach even in Reown's AppKit. Adding support for it is a different task.

Please review this so we can merge it on Tuesday. All the fixes and improvements will be done in another PR, but don't let this stop you commenting on the code.

### CODE REVIEW

File: Packages/io.chainsafe.web3-unity/Runtime/Scripts/Connection/ReownConnectionProvider.cs [hunk 1]
```diff
@@ -29,49 +33,34 @@ public class WalletConnectConnectionProvider : ConnectionProvider, IWalletConnec
...
-
-        [SerializeField] private List<string> disabledWallets;
-
-        [field: SerializeField] public WalletConnectLogLevel LogLevel { get; private set; } = WalletConnectLogLevel.ErrorOnly;
+        [field: SerializeField] public ReownLogLevel LogLevel { get; private set; } = ReownLogLevel.ErrorOnly;
 
         [field: SerializeField]
         public WalletLocationOption WalletLocationOption { get; private set; } = WalletLocationOption.LocalAndRemote;
 
-        [field: SerializeField, DefaultAssetValue("Packages/io.chainsafe.web3-unity/Runtime/Sprites/Logo_WalletConnect.png")]
```
Comment (kantagara) [inline Packages/io.chainsafe.web3-unity/Runtime/Scripts/Connection/ReownConnectionProvider.cs:48]: Are you missing the Sprite of the Reown? If so maybe still reuse the WC one if there is no decet Reown logo.

**Review by juans-chainsafe — CHANGES_REQUESTED** [review]
- Not an actual issue, but annoying, in the editor when you click a sample that requires to sign a transaction (like Mint), I get a window error "The application can't be opened" but the popup appears in my phone and I can sign the tx and see the result in the console (this may happen only in Mac, but it didn't happen with WC and is not happing with web3auth or Hyperplay)

Screenshot 2024-10-21 at 1 28 18 PM

- Wallets that appear in the editor list looks like they are for mobile, so they don't work in desktop

Screenshot 2024-10-21 at 1 31 56 PM

- Not working in mobile, if a select a wallet provider, like Metamask, I get this error in the console:
```
NullReferenceException: Object reference not set to an instance of an object.
  at ChainSafe.Gaming.Reown.RedirectionHandler.GetWalletAppDeeplink (ChainSafe.Gaming.Reown.Models.WalletModel walletData) [0x00000] in <00000000000000000000000000000000>:0 
  at ChainSafe.Gaming.Reown.RedirectionHandler.BuildConnectionDeeplink (ChainSafe.Gaming.Reown.Models.WalletModel walletData, System.String wcUri) [0x00000] in <00000000000000000000000000000000>:0 
  at ChainSafe.Gaming.Reown.RedirectionHandler.RedirectConnection (System.String connectionUri, System.String walletId) [0x00000] in <00000000000000000000000000000000>:0 
  at ChainSafe.Gaming.Reown.Dialog.ConnectionDialogBase.OnLocalWalletButtonClick (System.String walletName) [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.Events.UnityEvent.Invoke () [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.EventSystems.ExecuteEvents.Execute[T] (UnityEngine.GameObject target, UnityEngine.EventSystems.BaseEventData eventData, UnityEngine.EventSystems.ExecuteEvents+EventFunction`1[T1] functor) [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.EventSystems.StandaloneInputModule.ProcessTouchPress (UnityEngine.EventSystems.PointerEventData pointerEvent, System.Boolean pressed, System.Boolean released) [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.EventSystems.StandaloneInputModule.ProcessTouchEvents () [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.EventSystems.StandaloneInputModule.Process () [0x00000] in <00000000000000000000000000000000>:0 
--- End of stack trace from previous location where exception was thrown ---
```

File: Packages/io.chainsafe.web3-unity/Runtime/Scripts/Connection/ReownConnectionProvider.cs (hunk 1 above)
Comment (creeppak) [reply Packages/io.chainsafe.web3-unity/Runtime/Scripts/Connection/ReownConnectionProvider.cs:48]: Yes, I'm missing one. Gonna add it now. But I actually removed the attribute cause I remembered that you can set default values for UnityEngine.Object child classes inside the inspector when you select a script. It also binds them by GUID and not the path, so you can move the asset freely.

**Comment by creeppak** [comment]
> * Wallets that appear in the editor list looks like they are for mobile, so they don't work in desktop

For some reason we can't filter wallets for desktop now, there is only iOS and Android in their new API. For all other cases it's just gonna show all the wallets available even if there is no client for desktop.

But I think I found a solution, gonna filter out the wallets than don't have the desktop link.

**Comment by creeppak** [comment]
> * Not an actual issue, but annoying, in the editor when you click a sample that requires to sign a transaction (like Mint), I get a window error "The application can't be opened" but the popup appears in my phone and I can sign the tx and see the result in the console (this may happen only in Mac, but it didn't happen with WC and is not happing with web3auth or Hyperplay)

Hm.. I thought I fixes that one. So you connect with mobile, but it still tries to open something when a confirmation is expected from your wallet? Any other cases when it shows the error like that?

File: Packages/io.chainsafe.web3-unity/Runtime/Scripts/Connection/ReownConnectionProvider.cs (hunk 1 above)
Comment (creeppak) [reply Packages/io.chainsafe.web3-unity/Runtime/Scripts/Connection/ReownConnectionProvider.cs:48]: Added a new sprite for Reown.

File: src/ChainSafe.Gaming.Reown/Models/WalletModel.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    /// </summary>
+    public class WalletModel
+    {
+        [JsonProperty("id")] public string Id { get; set; }
+
+        [JsonProperty("name")] public string Name { get; set; }
+
+        [JsonProperty("homepage")] public string Homepage { get; set; }
+
+        [JsonProperty("image_id")] public string ImageId { get; set; }
```
Comment (rob1997) [inline src/ChainSafe.Gaming.Reown/Models/WalletModel.cs:16]: is this like image link?

File: src/ChainSafe.Gaming.Reown/Wallets/ReownWalletRegistry.cs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+            string BuildWalletIdsFilter(IList<string> walletIds)
+            {
+                return walletIds is { Count: > 0 }
+                    ? string.Join(",", walletIds)
+                    : null;
+            }
+        }
+
+        private static string BuildUriParameters(Dictionary<string, string> parameters)
```
Comment (rob1997) [inline src/ChainSafe.Gaming.Reown/Wallets/ReownWalletRegistry.cs:98]: can't we move this to `ReownHttpClient` more in context there imo

**Review by rob1997 — COMMENTED** [review]
Nice work!! left a few comments

File: Packages/io.chainsafe.web3-unity/Runtime/Scripts/Connection/ReownConnectionProvider.cs (hunk 1 above)
Comment (rob1997) [reply Packages/io.chainsafe.web3-unity/Runtime/Scripts/Connection/ReownConnectionProvider.cs:48]: > you can set default values for UnityEngine.Object child classes inside the inspector when you select a script. It also binds them by GUID and not the path, so you can move the asset freely.

How do you do that?

Comment (creeppak) [reply Packages/io.chainsafe.web3-unity/Runtime/Scripts/Connection/ReownConnectionProvider.cs:48]: > How do you do that?

Just select the *.cs file in Unity, you will see it in the inspector.

File: Packages/io.chainsafe.web3-unity/Runtime/Scripts/Connection/ReownConnectionProvider.cs [hunk 4]
```diff
@@ -1,24 +1,28 @@
...
     /// <summary>
-    /// WalletConnect connection provider used for connecting to a wallet using WalletConnect.
+    /// Reown connection provider used for connecting to a wallet using Reown.
     /// </summary>
-    [CreateAssetMenu(menuName = "ChainSafe/Connection Provider/Wallet Connect", fileName = nameof(WalletConnectConnectionProvider))]
-    public class WalletConnectConnectionProvider : ConnectionProvider, IWalletConnectConfig, IConnectionHandlerProvider
+    [CreateAssetMenu(menuName = "ChainSafe/Connection Provider/Reown", fileName = nameof(ReownConnectionProvider))]
+    public class ReownConnectionProvider : ConnectionProvider, IReownConfig, IConnectionHandlerProvider
     {
+        [SerializeField] private ConnectionHandlerBehaviour handlerPrefab;
```
[the remark this answers is not in the document]
Comment (creeppak) [reply Packages/io.chainsafe.web3-unity/Runtime/Scripts/Connection/ReownConnectionProvider.cs:22]: it's set in the inspector as I said. the GUID is stored in the meta file of the script.

File: Packages/io.chainsafe.web3-unity/Runtime/Scripts/Reown/AotConfiguration.cs [hunk 5]
```diff
@@ -26,7 +24,7 @@ void SetupAOT()
             Debug.Log(historyFactory.JsonRpcHistoryOfType<SessionExtend, Boolean>().GetType().FullName);
             Debug.Log(historyFactory.JsonRpcHistoryOfType<SessionDelete, Boolean>().GetType().FullName);
             Debug.Log(historyFactory.JsonRpcHistoryOfType<SessionPing, Boolean>().GetType().FullName);
-            EventManager<string, GenericEvent<string>>.InstanceOf(null).PropagateEvent(null, null);
+            // EventManager<string, GenericEvent<string>>.InstanceOf(null).PropagateEvent(null, null); todo there is no "Events" nuget package for Reown, this might break something
```
[the remark this answers is not in the document]
Comment (creeppak) [reply Packages/io.chainsafe.web3-unity/Runtime/Scripts/Reown/AotConfiguration.cs:27]: yes, we added this so it doesn't get stripped off. there is now no package and no notion of EventManager anywhere

File: scripts/data/published_dependencies.txt [hunk 6]
```diff
@@ -18,7 +18,6 @@ Packages/io.chainsafe.web3-unity/Runtime/Libraries/:
 	Nethereum.Web3
 	ChainSafe.Gaming.Unity
 	System.Buffers
-	ChainSafe.Gaming.WalletConnect
```
[the remark this answers is not in the document]
Comment (creeppak) [reply scripts/data/published_dependencies.txt:21]: np, I'm gonna duplicate this when merging dev

File: src/ChainSafe.Gaming.Reown/Models/WalletModel.cs (hunk 2 above)
Comment (creeppak) [reply src/ChainSafe.Gaming.Reown/Models/WalletModel.cs:16]: No, they use their new dedicated API now for all wallet-related data distribution. Below is the endpoint we should be sending requests to. It's improtant to use `ReownHttpClient` for this as it sets the required HTTP headers.

https://api.web3modal.com/getWalletImage/{wallet.ImageId}

File: src/ChainSafe.Gaming.Reown/Wallets/ReownWalletRegistry.cs (hunk 3 above)
Comment (creeppak) [reply src/ChainSafe.Gaming.Reown/Wallets/ReownWalletRegistry.cs:98]: I would actually even move this to an HttpUtil of some sort

Comment (creeppak) [reply src/ChainSafe.Gaming.Reown/Wallets/ReownWalletRegistry.cs]: Moved it out

File: Packages/io.chainsafe.web3-unity/Runtime/Scripts/Reown/Dialog/LocalWalletButton.cs [hunk 7]
```diff
@@ -28,20 +28,20 @@ public async void Set(WalletModel data, Action onClick)
...
-            var icon = await DownloadIcon(data.Images.SmallUrl);
-            if (icon == null)
-            {
-                return;
-            }
-
-            Icon.sprite = icon;
-            Icon.gameObject.SetActive(true);
-            IconStub.gameObject.SetActive(false);
+            // if (data.Images == null || string.IsNullOrEmpty(data.Images.SmallUrl)) // todo implement image downloading
```
[the remark this answers is not in the document]
Comment (creeppak) [reply Packages/io.chainsafe.web3-unity/Runtime/Scripts/Reown/Dialog/LocalWalletButton.cs:31]: yeaaap

File: Packages/io.chainsafe.web3-unity/Runtime/Scripts/Connection/ReownConnectionProvider.cs (hunk 1 above)
Comment (rob1997) [reply Packages/io.chainsafe.web3-unity/Runtime/Scripts/Connection/ReownConnectionProvider.cs:48]: I had no idea xD, thanks!

File: Packages/io.chainsafe.web3-unity/Runtime/Scripts/Reown/AotConfiguration.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (rob1997) [reply Packages/io.chainsafe.web3-unity/Runtime/Scripts/Reown/AotConfiguration.cs:27]: Does it still build to those platforms? I think it was IOS that was failing, right?

**Review by juans-chainsafe — CHANGES_REQUESTED** [review]
The previous 3 issues were fixed!

I have 2 more things at the moment (I still need to try with mainnet in other wallet providers to see how it goes in iOS)

- In desktop, I only see these 6 providers, before we could scroll down a lot more, is this on purpose?
Screenshot 2024-10-22 at 9 46 52 AM

- In mobile, after you connect Metamask and you click a sample that requires to sign a transaction, you get redirected to Metamask, sign the transaction, go back to the Unity app and the loading spinner is there forever, even when the transaction finishes. And it also happens with simple things like Sign Message sample (let me know if you need a video for this)

File: Packages/io.chainsafe.web3-unity/Runtime/Scripts/Reown/AotConfiguration.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (creeppak) [reply Packages/io.chainsafe.web3-unity/Runtime/Scripts/Reown/AotConfiguration.cs:27]: Haven't checked it yet TBH.

**Comment by creeppak** [comment]
> * In desktop, I only see these 6 providers, before we could scroll down a lot more, is this on purpose?
> 
> Screenshot 2024-10-22 at 9 46 52 AM

This is the only wallets supported by Reown that are available on desktop. I guess all the old ones we had didn't work anyway, so now we only show those that are 'officialy' supported by Reown on desktop.

**Comment by creeppak** [comment]
> * In mobile, after you connect Metamask and you click a sample that requires to sign a transaction, you get redirected to Metamask, sign the transaction, go back to the Unity app and the loading spinner is there forever, even when the transaction finishes. And it also happens with simple things like Sign Message sample (let me know if you need a video for this)

That's weird, never experienced anything similar. Could you please create an issue with steps to reproduce?

**Review by kantagara — APPROVED** [review]

**Comment by juans-chainsafe** [comment]
> > * In mobile, after you connect Metamask and you click a sample that requires to sign a transaction, you get redirected to Metamask, sign the transaction, go back to the Unity app and the loading spinner is there forever, even when the transaction finishes. And it also happens with simple things like Sign Message sample (let me know if you need a video for this)
> 
> That's weird, never experienced anything similar. Could you please create an issue with steps to reproduce?

https://github.com/ChainSafe/web3.unity/issues/1201

**Review by rob1997 — APPROVED** [review]

**Review by juans-chainsafe — APPROVED** [review]
The current status of the wallets that you can connect is awesome!
 connect wallet status in iOS:
Metamask ✅ 
Trust Wallet ✅ 
Zerion ✅ 
Ledger - I can’t connect it right now, my ledger is not configured in the iphone
Rainbow Wallet ✅ 
Exodus ✅ 
MyEtherWallet - doesnt have a mobile app
ZenGo - is not sending me the email to login, so I can’t test it right now
Phantom Wallet ❌  Doesn’t appear in the WC list
Coinbase wallet ❌  is not on the list

**Review by kantagara — APPROVED** [review]

**Pull request merged**